### PR TITLE
Do trace storage allocation on the Rust side.

### PR DESCRIPTION
**Pull request by vext01**
This change ensures that Rust does both the `malloc(3)` and `free(3)` of the trace buffers.

This was hard, but I think I got there in the end. valgrind happy, all tests pass.

Looks good?

### CODE REVIEW

File: src/backends/perf_pt/mod.rs [hunk 1]
```diff
@@ -260,10 +278,18 @@ impl Tracer for PerfPTTracer {
             return Err(HWTracerError::TracerAlreadyStarted);
         }
 
-        if unsafe { perf_pt_start_tracer(self.tracer_ctx) } == -1 {
+        // It is essential we box the trace now to stop it from moving. If it were to move, then
+        // the reference which we pass to C here would become invalid. The interface to
```
Comment (ltratt) [inline src/backends/perf_pt/mod.rs]: Is this statement true? I assume you're not passing a poiunter to `PerfPTTrace`, at which point it doesn't matter if `PerfPTTrace` moves, so long as the C memory inside it doesn't?

File: src/backends/perf_pt/perf_pt.c [hunk 2]
```diff
@@ -400,14 +406,20 @@ perf_pt_init_tracer(struct tracer_conf *tr_conf)
...
  * Returns zero on success, or -1 otherwise.
  */
 int
-perf_pt_start_tracer(struct tracer_ctx *tr_ctx)
+perf_pt_start_tracer(struct tracer_ctx *tr_ctx, struct perf_pt_trace *trace)
 {
     int clean_sem = 0, clean_thread = 0;
     int ret = 0;
 
+
```
Comment (ltratt) [inline src/backends/perf_pt/perf_pt.c]: Double blank line?

File: src/backends/perf_pt/mod.rs (hunk 1 above)
Comment (vext01) [reply src/backends/perf_pt/mod.rs]: We are passing a pointer to a `PerfPTTrace`, as C needs to access the other fields too. 

Even if we were only passing the buffer itself, we would need to pass a `*mut *mut u8` which would have the same problem. Although the underlying C-level heap allocation won't move, the intermediate pointer might. The intermediate pointer is needed for a `realloc(3)` when we get around to implementing #33.

[I actually found this problem *before* i was passing a reference to the whole struct]

File: src/backends/perf_pt/perf_pt.c (hunk 2 above)
Comment (vext01) [reply src/backends/perf_pt/perf_pt.c]: Oops, yes. That'll be a debugging leftover.

**Comment by vext01** [comment]
Comments addressed.

File: src/backends/perf_pt/mod.rs (hunk 1 above)
Comment (ltratt) [reply src/backends/perf_pt/mod.rs]: OK, I think I get it.

**Comment by ltratt** [comment]
If you squash, I can merge.

**Comment by vext01** [comment]
splat

File: src/backends/perf_pt/mod.rs (hunk 1 above)
Comment (vext01) [reply src/backends/perf_pt/mod.rs]: It's a mind bender, I agree. It took me most part of a day to find this bug!

**Pull request merged**